### PR TITLE
Migrate the tax rate management screens to Backbone/Underscore.

### LINKED ISSUE #8587: Search tax rate
Could be useful to have a search form in the tax rates table page.
When you have a lot of different tax rates it's hard to find the one you are looking for.

### PR DESCRIPTION
**Pull request by georgestephanis**
Fixes #8587 -- and a whole lot more.
- Enabled search!
- Separates out the JS from being inline, meaning that it can be properly scanned with `jshint` and cached!
- Redoes how pagination is performed, so Search and Pagination can run purely in JS.
- When saving tax rates, it now only pushes up to the server the fields that have changed.  This can be critical for servers that have a `max_input_vars` limit set either within php or suhosin -- default (if enabled) is 1,000 fields, and with an old limit of 100 per page, that could easily get passed.
- When editing form fields, it sets a flag that data on the page has changed, so if you go to navigate away and lose your changes, it prompts you to save it instead.  Yay fewer lost changes.
- When searching, sorting is temporarily disabled.
- All row templating is now done in one place -- JS -- rather than by PHP originally and then recreated in JS for adding new rows and such.
- Got DB queries out of the template file.
- Probably a number of other small things that don't immediately come to mind.

### CODE REVIEW

File: includes/admin/settings/views/html-settings-tax.php [hunk 1]
```diff
@@ -3,273 +3,137 @@
...
-		});
+<div id="unsaved-changes" style="display:none;">
+	<?php _e( 'There are unsaved changes.', 'woocommerce' ); ?>
+	<pre style="display:none;"></pre>
+</div>
+
+<script type="text/html" id="tmpl-wc-tax-table-row">
+	<tr class="tips" data-tip="<?php echo esc_attr( sprintf( __( 'Tax rate ID: %s', 'woocommerce' ), '{{ data.tax_rate_id }}' ) ); ?>" data-id="{{ data.tax_rate_id }}">
+		<td class="sort">
+			<# if ( ! data.newRow ) { #>
```
Comment (vestimir) [inline includes/admin/settings/views/html-settings-tax.php]: Is this really needed?

**Comment by vestimir** [comment]
General JS note → You might want to split the big javascript file into several smaller (one for the model, one for the view controller, etc.) and extract logic out of it (for example the export, or the calculations in add/delete methods). I think it'd be more readable and easy to support.

File: includes/admin/settings/views/html-settings-tax.php (hunk 1 above)
Comment (georgestephanis) [inline includes/admin/settings/views/html-settings-tax.php]: It was to mimic the old structure of how it used to track additions/removals.  It can probably go away, I was just planning on doing a second pull request to remove the no-longer-used functions from the old method, once the new is merged in.

**Comment by georgestephanis** [comment]
@vestimir As they're only ever going to be used in conjunction with one another, I'm not sure I understand the benefit -- the downside, of course, is three HTTP requests over one.

The logic I had just migrated into the view, as it made sense to me that it would be dependent on the table that the view is populating into.  So, given the table and the model, it can attach everything.

Not that I'm locked into that pattern by any means, I just don't understand the reason.

**Comment by vestimir** [comment]
@georgestephanis I see, so it makes more sense to keep it in one file. 

And for the big Backbone view, my motivation is that it's hard to follow what's going on here with so much code gone into one entity. Extracting logic that calculates stuff, or formats stuff would make it more readable, and also remove duplication like the one for `rates_to_reorder` and `reordered_rates` in `onAddNewRow`/`onDeleteRow`. 

Another example is the `onExport` method which holds logic that's more close to the data, than to the view controller itself. I'd go with a separate `TaxCsvExport` class that does the stuff contained within this method and just return its result. Something along:

```js
var TaxCsvExport = {
    getCSV: function (rates) {
       var csv_data = 'data:application/csv;charset=utf-8,' + data.strings.csv_data_cols.join(',') + '\n';

       $.each( rates, function( id, rowData ) {
           var row = '';

           row += rowData.tax_rate_country  + ',';
           row += rowData.tax_rate_state    + ',';
           row += ( rowData.postcode        ? rowData.postcode.join( '; ' ) : '' ) + ',';
           row += ( rowData.city            ? rowData.city.join( '; ' )     : '' ) + ',';
           row += rowData.tax_rate          + ',';
           row += rowData.tax_rate_name     + ',';
           row += rowData.tax_rate_priority + ',';
           row += rowData.tax_rate_compound + ',';
           row += rowData.tax_rate_shipping + ',';
           row += data.current_class;

           csv_data += row + '\n';
       });
    }
};

// And in the view class
onExport : function( event ) {
   $(this).attr( 'href', encodeURI( TaxCsvExport.getCSV(event.data.view.model.rates) ) );

   return true;
},
```

Although it introduces a new entity to the whole story, such change improves the overall readability and also makes it easier to refactor the new object (for example switch `$.each` to `_.reduce` and the `row` to an Array that's joined with `,` in the end).

**Comment by mikejolley** [comment]
@gedex Triggers on insert:

> settings-views-html-settings-tax.js?ver=2.4.3:197 Uncaught TypeError: Cannot read property 'tax_rate_order' of undefined

Also the search input needs styling/aligning.
